apiserver/facades/controller/caasoperatorprovisioner/provisioner.go

```diff
@@ -57,3 +66,133 @@ func (a *API) WatchApplications() (params.StringsWatchResult, error) {
 	}
 	return params.StringsWatchResult{}, watcher.EnsureErr(watch)
 }
+
+// UpdateApplicationsUnits updates the Juju data model to reflect the given
+// units of the specified application.
+func (a *API) UpdateApplicationsUnits(args params.UpdateApplicationUnitArgs) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Args)),
+	}
+	if len(args.Args) == 0 {
+		return result, nil
+	}
+	for i, appUpdate := range args.Args {
+		appTag, err := names.ParseApplicationTag(appUpdate.ApplicationTag)
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		app, err := a.state.Application(appTag.Id())
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		if err := a.updateUnitsFromCloud(app, appUpdate.Units); err != nil {
+			result.Results[i].Error = common.ServerError(err)
+		}
+	}
+	return result, nil
+}
+
+// UpdateUnitsFromCloud takes a slice of unit information provided by an external
+// source (typically a cloud update event) and merges that with the existing unit
+// data model in state.
```

Comment:
maybe add to the comment stating that the set of units provided is *complete* for that external source, and that any units with provider IDs in state that aren't in the set will be removed

Reply:
ok